### PR TITLE
bpo-35983: skip trashcan for subclasses

### PR DESCRIPTION
**Pull request by jdemeyer**
https://bugs.python.org/issue35983

### CODE REVIEW

**Review by 1st1 — APPROVED** [review]

**Comment by 1st1** [comment]
LGTM, but let another core dev review this before merge.  @pitrou could you please take a look?

**Comment by jdemeyer** [comment]
Please consider backporting this to 2.7 and 3.7

**Comment by jdemeyer** [comment]
Any news?

**Comment by pitrou** [comment]
I was asked for review. Unfortunately this is a hairy topic and I would have to take a dive to make a proper review. Depending how confident other core developers are, this could be merged anyway. @1st1, @serhiy-storchaka what do you think?

@scoder perhaps you'd like to take a look?

**Comment by pitrou** [comment]
(PS: one issue with the trashcan mechanism is that it triggers rarely, so any bugs may occur only in obscure conditions that we don't have access to...)

**Comment by scoder** [comment]
I'm absolutely not an expert in this area, but several things speak in favour of Jeroen's change:
1) He found a case that currently does not work and creates user visible problems.
2) He found code in CPython that already tries to work around this for a specific case.
3) He managed to implement a test for this.
4) He came up with an implementation that seems to use backwards compatible macros.
5) Looking over the change, neither @1st1 nor I (and apparently also not @pitrou) could find anything obviously wrong with it, even though no-one of us felt confident enough to determine that it's definitely correct.
6) The current implementation obviously is less correct, given that it has at least one bug, which is not currently covered by tests.

My vote is for merging it and allowing users to find problems during the alpha/beta cycle. Several major projects have integration builds with the latest CPython master branch these days.

File: Include/object.h
Comment (pitrou) [inline Include/object.h:719]: Out of consistency and for future proofness, I think this should still be `Py_TRASHCAN_END(op)`.

File: Modules/_testcapimodule.c
Comment (pitrou) [inline Modules/_testcapimodule.c]: Use `Py_FatalError` instead?

File: Lib/test/test_capi.py
Comment (pitrou) [inline Lib/test/test_capi.py:369]: Is `2**20` really required to exercise this?

File: Objects/odictobject.c
Comment (pitrou) [inline Objects/odictobject.c:1360]: Is there a trashcan test for `OrderedDict`? If not, perhaps add one?

Comment (pitrou) [reply Objects/odictobject.c:1360]: (or at least validate offline that it works)

File: Objects/odictobject.c
Comment (pitrou) [inline Objects/odictobject.c:1367]: By the way, why is it not needed to retrack the object here?

**Review by pitrou — CHANGES_REQUESTED** [review]
Here is a review after all.

File: Lib/test/test_capi.py
Comment (jdemeyer) [reply Lib/test/test_capi.py:369]: > Is `2**20` really required to exercise this?

Yes: a typical stack on Linux is 8MB, so you need to use `2**23 / S` iterations, where `S` is the size of a stack frame. As bare minimum, the stack frame needs to contain the return address, so that's 8 bytes on a 64-bit system. I think it's possible to actually achieve this minimum, when omitting the frame pointer and storing all variables in registers.

File: Include/object.h
Comment (jdemeyer) [reply Include/object.h:719]: > Out of consistency

If you really want to be consistent, it should be `Py_TRASHCAN_END(op, dealloc)`. But I don't see the need for such "consistency".

> and for future proofness

I don't think it's needed for future proofness either: if we ever require `op` in the `Py_TRASHCAN_END` macro, it would be better to store it in a temporary variable and use that instead. That also avoids potential errors where somebody plugs in different values for `op` in the BEGIN and END macros.

But in the end, this is just bikeshedding. So if you insist, I will add the useless `op` argument.

File: Objects/odictobject.c
Comment (jdemeyer) [reply Objects/odictobject.c:1367]: > By the way, why is it not needed to retrack the object here?

Because there is no general requirement to retrack. A typical `tp_dealloc` implementation uses `PyObject_GC_UnTrack()` which is a no-op if the object is already untracked. Retracking is only needed if the `tp_dealloc` of the base class uses unsafe GC macros that might crash on non-GC-tracked objects.

Conclusion: in the case where we don't know the base class, we need to retrack. If we know that the base class is fine, there is no need.

File: Include/object.h
Comment (pitrou) [reply Include/object.h:719]: Fair enough.

File: Objects/odictobject.c
Comment (scoder) [reply Objects/odictobject.c:1367]: This sounds like something to carve into a stone somewhere.

Comment (jdemeyer) [reply Objects/odictobject.c:1367]: It's already documented in `Objects/typeobject.c`:
```
       Q. Why do we GC-untrack before the trashcan and then immediately
          GC-track again afterward?

       A. In the case that the base class is GC-aware, the base class
          probably GC-untracks the object.      If it does that using the
          UNTRACK macro, this will crash when the object is already
          untracked.  Because we don't know what the base class does, the
          only safe thing is to make sure the object is tracked when we
          call the base class dealloc.  But...  The trashcan begin macro
          requires that the object is *untracked* before it is called.  So
          the dance becomes:

         GC untrack
         trashcan begin
         GC track
```
As far as I can tell, this is the only place in CPython where we do retrack before calling the base `tp_dealloc`.

File: Objects/odictobject.c
Comment (jdemeyer) [reply Objects/odictobject.c:1360]: > Is there a trashcan test for `OrderedDict`?

Yes, added by https://bugs.python.org/issue25395

**Review by pitrou — APPROVED** [review]

File: Include/object.h
Comment (serhiy-storchaka) [inline Include/object.h:714]: Why not use `else`?

Comment (jdemeyer) [reply Include/object.h:714]: Because of the particular code structure (a `break` inside a nested `if`), I don't see a beautiful/simple way to make it work with `else`. This is a good example where a `goto`-like statement (in this case, a `break`) works best.

**Comment by jdemeyer** [comment]
If you all agree, please remove the `DO-NOT-MERGE` label.

**Comment by pitrou** [comment]
@serhiy-storchaka is the last one that needs approving.

**Comment by pitrou** [comment]
Btw, if we are confident enough this could be a backport candidate.

**Review by methane — APPROVED** [review]

**Comment by jdemeyer** [comment]
By https://discuss.python.org/t/vote-to-promote-stefan-behnel-as-a-core-developer/1054 it looks like this will have positive reviews from no less than 4 core developers.

So @serhiy-storchaka could you please say more concretely why you do not approve?

**Comment by jdemeyer** [comment]
> Btw, if we are confident enough this could be a backport candidate.

I'm seeing actual crashes caused by this on Python 2.7, so I'm interested in backporting to 2.7. The patch doesn't apply cleanly on top of 2.7, but I'm willing to work on that.

**Comment by pitrou** [comment]
@serhiy-storchaka If you don't mind, I will merge this soon.

**Comment by pitrou** [comment]
I think @benjaminp will have to decide whether this is 2.7-backport-worthy or not.

File: Include/object.h
Comment (serhiy-storchaka) [inline Include/object.h:669]: Why Py_TRASHCAN_SAFE_END has been renamed to Py_TRASHCAN_END? Is it not safe anymore?

File: Lib/test/test_capi.py
Comment (serhiy-storchaka) [inline Lib/test/test_capi.py:336]: Could you add a test for a Python subclass of MyList?

File: Include/object.h
Comment (jdemeyer) [reply Include/object.h:669]: > Why Py_TRASHCAN_SAFE_END has been renamed to Py_TRASHCAN_END? Is it not safe anymore?

I changed the signature, so I had to change the name too. `Py_TRASHCAN_BEGIN` and `Py_TRASHCAN_END` then seemed like the most logical names.

Comment (serhiy-storchaka) [reply Include/object.h:669]: Could not you preserve the signature of Py_TRASHCAN_SAFE_END? This would minimize changes.

Having two sets of names with and without SAFE gives false hint that the SAFE pair is more preferable.

Comment (jdemeyer) [reply Include/object.h:669]: > Could not you preserve the signature of Py_TRASHCAN_SAFE_END?

Of course I could, but what's the point of passing an argument which is not used?

> This would minimize changes.

Why is that a goal? My goal is producing a good and simple trashcan implementation, not to minimize changes.

> Having two sets of names with and without SAFE gives false hint that the SAFE pair is more preferable.

When reading the comments in the code (the only place where these are documented), it's clear that one should use `Py_TRASHCAN_BEGIN` and that `Py_TRASHCAN_SAFE_BEGIN` is only meant for backwards compatibility.

This is all bikeshedding of course and I don't really care much about the name and the unused arguments. So, if you have a concrete proposal that you're willing to accept, just say it.

**Comment by jdemeyer** [comment]
ping?

**Comment by pitrou** [comment]
I'll let @serhiy-storchaka answer your last questions.

**Comment by pitrou** [comment]
There's no need to hold this up any longer. I'm merging.

**Comment by pitrou** [comment]
Thanks for being persistent @jdemeyer :-)

**Pull request merged**
